include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -99,6 +99,36 @@ ORT_DEFINE_RELEASE(ModelMetadata);
 ORT_DEFINE_RELEASE(ThreadingOptions);
 ORT_DEFINE_RELEASE(IoBinding);
 
+/// Ort::Float16_t/BFloat16_t are a structures that represents float16/bfloat16 data.
+/// They are necessary for type dispatching to work within C++ API so one can feed/fetch
+/// data via C++ API.
+/// Both types are implicitly convertible to/from uint16_t.
+/// The size of the structs should align with uint16_t and one can freely cast
+/// uint16_t buffers to/from Ort::Float16/BFloat16 to feed and retrieve data.
+struct Float16_t {
+  uint16_t v_;
+  constexpr Float16_t() noexcept : v_(0) {}
+  constexpr Float16_t(uint16_t v) noexcept : v_(v) {}
+  constexpr operator uint16_t() const noexcept { return v_; }
+  constexpr uint16_t get() const noexcept { return v_;  }
+  constexpr bool operator==(Float16_t rhs) const noexcept { return v_ == rhs.v_; };
+  constexpr bool operator!=(Float16_t rhs) const noexcept { return v_ != rhs.v_; };
```

Comment:
some interface questions
1. the name "get" seems a bit vague here, and do we need it in addition to operator uint16_t()/directly accessing v_?
2. why are only operator== and operator!= implemented and not the other relational operators?
3. should v_ be private? if it is private, maybe operator=(uint16_t) would be useful too

Reply:
1. get() is a standard way of getting things w/o a cast and makes the intent more explicit.
2. Good point
3. May be, but that would make it more like a full fledged class rather than POD which it should be. But OK.

---
In reply to: [518461183](https://github.com/microsoft/onnxruntime/pull/5716#discussion_r518461183) [](ancestors = 518461183)